_CS141/type-level-programming.md

```diff
@@ -157,16 +157,92 @@ stwo :: SNat ('Succ ('Succ 'Zero))
 stwo :: SSucc sone
 ```
 
-## Proxy Types
+## Reification
 
-We don’t have types at runtime due to **type erasure**, so we can’t write a function that does something based on a type. 
+> Reification is the process of **converting types to values**, typically using **type classes**. This resolves to the correct instances at compile-time so the right implementation of the function is evaluated at runtime.
 
-> Proxy types allow us work around Haskell’s limitations and convert a representation of a value at the **type level** into a **term level** value. 
+There are a few ways to implement the instances for the type classes and one way (which is used in the lectures) is to use **proxy types**.
+
+### Proxy Types
+
+We don’t have types at runtime due to **type erasure**, so we can’t write a function that takes types as arguments in Haskell, even though we sometimes want to.
```

Comment:
```suggestion
We don’t have types at runtime due to **type erasure**, so we can’t write a function that takes types as arguments in Haskell, even though we sometimes want to. For example, we might want to define a function `fromNat :: forall (n :: Nat) => n -> Int` which takes a type-level natural number `n` of kind `Nat` as input and returns the corresponding `Int` value. However, only types of kind `*` have values at run-time in Haskell. Since there are no values of types of kind `Nat`, we cannot define a `fromNat` function of the given type.
```